browser/ui/webui/new_tab_page/ms_edge_protocol_message_handler.cc

```diff
@@ -0,0 +1,128 @@
+
+#include "brave/browser/ui/webui/new_tab_page/ms_edge_protocol_message_handler.h"
+
+#include "base/bind.h"
+#include "base/command_line.h"
+#include "base/files/file_path.h"
+#include "base/logging.h"
+#include "base/notreached.h"
+#include "base/path_service.h"
+#include "base/task/thread_pool.h"
+#include "brave/browser/default_protocol_handler_utils_win.h"
+#include "chrome/installer/util/shell_util.h"
+
+namespace {
+
+constexpr wchar_t kMSEdgeProtocol[] = L"microsoft-edge";
+constexpr wchar_t kMSEdgeProtocolRegKey[] =
+    L"SOFTWARE\\Microsoft\\Windows\\Shell\\Associations\\UrlAssociations\\micro"
+    L"soft-edge";
+
+}  // namespace
+
+MSEdgeProtocolMessageHandler::MSEdgeProtocolMessageHandler()
+    : user_choice_key_(HKEY_CURRENT_USER, kMSEdgeProtocolRegKey, KEY_NOTIFY) {
+  StartWatching();
+}
+
+MSEdgeProtocolMessageHandler::~MSEdgeProtocolMessageHandler() = default;
+
+void MSEdgeProtocolMessageHandler::StartWatching() {
+  if (user_choice_key_.Valid()) {
+    user_choice_key_.StartWatching(
+        base::BindOnce(&MSEdgeProtocolMessageHandler::OnRegValChanged,
+                       base::Unretained(this)));
+  }
+}
+
+void MSEdgeProtocolMessageHandler::OnRegValChanged() {
+  CheckMSEdgeProtocolDefaultHandlerState();
+  StartWatching();
+}
+
+void MSEdgeProtocolMessageHandler::CheckMSEdgeProtocolDefaultHandlerState() {
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(&IsDefaultProtocolHandlerFor, kMSEdgeProtocol),
+      base::BindOnce(&MSEdgeProtocolMessageHandler::OnIsDefaultProtocolHandler,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void MSEdgeProtocolMessageHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "checkDefaultMSEdgeProtocolHandlerState",
+      base::BindRepeating(&MSEdgeProtocolMessageHandler::
+                              HandleCheckDefaultMSEdgeProtocolHandlerState,
+                          base::Unretained(this)));
+
+  web_ui()->RegisterMessageCallback(
+      "setAsDefaultMSEdgeProtocolHandler",
+      base::BindRepeating(&MSEdgeProtocolMessageHandler::
+                              HandleSetAsDefaultMSEdgeProtocolHandler,
+                          base::Unretained(this)));
+}
+
+void MSEdgeProtocolMessageHandler::HandleCheckDefaultMSEdgeProtocolHandlerState(
+    base::Value::ConstListView args) {
+  AllowJavascript();
+  CheckMSEdgeProtocolDefaultHandlerState();
+}
+
+void MSEdgeProtocolMessageHandler::HandleSetAsDefaultMSEdgeProtocolHandler(
+    base::Value::ConstListView args) {
+  AllowJavascript();
+
+  // Test purpose switch to use system ui.
+  constexpr char kUseSystemUIForMSEdgeProtocol[] = "use-system-ui-for-ms-edge";
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch(
+          kUseSystemUIForMSEdgeProtocol)) {
+    LaunchSystemDialog();
+    return;
+  }
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
```

Comment:
here it might be better to use `base::TaskPriority::USER_BLOCKING` (not USER_VISIBLE).